circuit-retained/src/commonMain/kotlin/com/slack/circuit/retained/RetainedStateHolder.kt

```diff
@@ -0,0 +1,79 @@
+// Copyright (C) 2024 Slack Technologies, LLC
+// SPDX-License-Identifier: Apache-2.0
+package com.slack.circuit.retained
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.CompositionLocalProvider
+import androidx.compose.runtime.DisposableEffect
+
+/**
+ * A holder that provides a unique retainedStateRegistry for each subtree and retains all preserved
+ * values. Each [RetainedStateProvider] maintains a unique retainedStateRegistry for each key,
+ * allowing it to save and restore states.
+ */
+public interface RetainedStateHolder {
+
+  /**
+   * Provides a [RetainedStateRegistry] for the child [content] based on the specified [key]. Before
+   * the provided registry is disposed, it calls [RetainedStateRegistry.saveValue] on the holder's
+   * registry to save the current value, allowing it to be restored on the next visit with the same
+   * key.
+   */
+  @Composable public fun RetainedStateProvider(key: String, content: @Composable () -> Unit)
+}
+
+/** Creates and remembers the instance of [RetainedStateHolder]. */
+@Composable
+public fun rememberRetainedStateHolder(): RetainedStateHolder {
+  return rememberRetained { RetainedStateHolderImpl() }
+}
+
+private class RetainedStateHolderImpl : RetainedStateHolder, RetainedStateRegistry {
+
+  private val registry = RetainedStateRegistry()
+
+  @Composable
+  override fun RetainedStateProvider(key: String, content: @Composable (() -> Unit)) {
+    CompositionLocalProvider(LocalRetainedStateRegistry provides this) {
+      val canRetainChecker = LocalCanRetainChecker.current ?: CanRetainChecker.Always
+      val childRegistry = rememberRetained(key = key) { RetainedStateRegistry() }
+      CompositionLocalProvider(
+        LocalRetainedStateRegistry provides childRegistry,
+        LocalCanRetainChecker provides CanRetainChecker.Always,
```

Comment:
When creating separate `RetainedStateRegistry` instances within `RetainedStateProvider`, the `CanRetainChecker` propagated to the child has a different meaning from the parent’s `CanRetainChecker`.

**Meaning of the child’s `CanRetainChecker`:**

The child’s `CanRetainChecker` determines whether `rememberRetained` declared under the `RetainedStateProvider` can be retained in cases where the key transitions from A -> B -> A. This behavior should always function correctly even if the parent’s `CanRetainChecker` always returns false, because the `RetainedStateHolder` itself has never been disposed. As long as the `RetainedStateHolder` is not disposed, `rememberRetained` declared under the `RetainedStateProvider` should always be restorable, so it seems appropriate to pass `CanRetainChecker.Always`.

**Meaning of the parent’s `CanRetainChecker`:**

The `CanRetainChecker` provided by the parent of the `RetainedStateHolder` determines whether the internal data of the `RetainedStateHolder` should be retained when the screen is recreated.

-----------

For these reasons, I think the existing code also passed `CanRetainChecker.Always` when creating a new registry. Following the intention of the original code, I believe redefining it as `CanRetainChecker.Always` in my code is necessary.

> I tried removing it, and all of the tests still pass, so maybe this can now be removed?

I think the current test cases might be insufficient. I expect a difference in behavior for the following scenario:

1. The parent’s CanRetainChecker always returns false.
2. A RetainedStateHolder is defined.
3. The key in holder.RetainedStateProvider transitions from A -> B -> A.

When redefined with CanRetainChecker.Always -> rememberRetained inside holder.RetainedStateProvider is preserved.

Without redefined ->
rememberRetained inside holder.RetainedStateProvider is not preserved.